doc/architectural/memory-primitives.md

```diff
@@ -91,6 +91,41 @@ point into another memory object (such as could happen when running on a real
 machine). To verify that pointers stay within the bounds of their pointees, the
 CBMC option `--pointer-overflow-check` can be used.
 
+### Malloc modelling
+
+CBMC ships a model of `malloc` that seeks to emulate the behaviour of the C
+standard library. By default, however, it will always succeed to allocate an
+object. This extends to the case of `malloc(0)`, which returns a valid pointer
+to an object. The size of that object is zero, implying that any attempt to read
+from or write to this object will result in an out-of-bounds access.
+
+The following command line options add emulation of failures:
+
+|Flag                    |  Check                                          |
+|------------------------|-------------------------------------------------|
+| `--malloc-fail-null`   |  in case malloc fails return NULL               |
+| `--malloc-fail-assert` |  in case malloc fails report as failed property |
+| `--malloc-may-fail`    |  malloc may non-deterministically fail          |
+
+Calling `malloc` may fail for a number of reasons and the function may return a
+NULL pointer. The users can choose if and how they want the `malloc`-related
+failures to occur. The option `--malloc-fail-null` results in `malloc` returning
+the NULL pointer when failing. The option `--malloc-fail-assert` places
+additional properties inside `malloc` that are checked and if failing the
+verification is terminated (by `assume(false)`). One such property is that the
```

Comment:
I'd drop the `assume(false)` -- this is an implementation detail.

Reply:
Instead give examples of user-code, one with `if(result == 0)` and one without, and say which option is appropriate for which.

Reply:
Done.